src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1310 @@
+            //       ├──▌  ASG       int
+            //       │  ├──▌  LCL_VAR   int    V35 tmp31
+            //       │  └──▌  LCL_FLD   int    V03 loc1         [+4]
+            // This really ought to be handled by local copy prop, but the way it works during
+            // morph makes it hard to fix there.
+            //
+            // This is the short term fix. Long term fixes may be:
+            // 1. Fix local copy prop
+            // 2. Teach LSRA to allow the above cases, simplifying IR concepts (e.g.
+            //    introduce something like GT_COPY on top of LCL_VAR when they
+            //    need to be "defs")
+            // 3. Change the pass here to avoid creating any embedded assignments by making use
+            //    of gtSplitTree. We will only need to split in very edge cases since the point
+            //    at which the replacement was marked as needing read back is practically always
+            //    going to be in a previous statement, so this shouldn't be too bad for CQ.
+
+            m_compiler->lvaGetDesc(rep.LclNum)->lvRedefinedInEmbeddedStatement = true;
+        }
+
+        m_madeChanges = true;
+    }
+
+    //------------------------------------------------------------------------
+    // StoreBeforeReturn:
+    //   Handle a return of a potential struct local.
+    //
+    // Parameters:
+    //   ret - The GT_RETURN node
+    //
+    void StoreBeforeReturn(GenTreeUnOp* ret)
+    {
+        if (ret->TypeIs(TYP_VOID) || !ret->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+        {
+            return;
+        }
+
+        GenTreeLclVarCommon* retLcl = ret->gtGetOp1()->AsLclVarCommon();
+        if (retLcl->TypeIs(TYP_STRUCT))
+        {
+            unsigned size = retLcl->GetLayout(m_compiler)->GetSize();
+            WriteBackBefore(&ret->gtOp1, retLcl->GetLclNum(), retLcl->GetLclOffs(), size);
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // WriteBackBefore:
+    //   Update the use with IR that writes back all necessary overlapping
+    //   replacements into a struct local.
+    //
+    // Parameters:
+    //   use  - The use, which will be updated with a cascasing comma trees of assignments
+    //   lcl  - The struct local
+    //   offs - The starting offset into the struct local of the overlapping range to write back to
+    //   size - The size of the overlapping range
+    //
+    void WriteBackBefore(GenTree** use, unsigned lcl, unsigned offs, unsigned size)
+    {
+        if (m_replacements[lcl] == nullptr)
+        {
+            return;
+        }
+
+        jitstd::vector<Replacement>& replacements = *m_replacements[lcl];
+        size_t                       index        = BinarySearch<Replacement, &Replacement::Offset>(replacements, offs);
+
+        if ((ssize_t)index < 0)
+        {
+            index = ~index;
+            if ((index > 0) && replacements[index - 1].Overlaps(offs, size))
```

Comment:
How do you know that there aren't any lower indexed entries that overlap?

Reply:
`replacements` has only disjoint ranges in it (ensured by `EvaluateReplacement`)